### PR TITLE
perf(encoding): #111 Phase 2 — saturating_* cleanup on hot path

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Phase 2 of the #111 encoder architecture rewrite — **saturating_* cleanup
on the hot path**.

The original Phase 2 plan called for both a `Cwksp` arena allocator and
a `saturating_*` cleanup. After implementing the cleanup and benchmarking,
the arena work was **deferred** to a measurement-driven follow-up (issue
#120 updated accordingly): Rust's `Vec`-based scratch already amortizes
per-frame allocation across frames, so the donor's "single cwksp + cursor
reset" model doesn't translate to a measurable win without `unsafe`
pointer juggling for disjoint `&mut [T]` slices.

## Architectural mechanism — frame-level overflow gate

A naive removal of `saturating_*` on `abs_pos + N` is unsafe on 32-bit
streaming encodes: the absolute stream cursor grows monotonically with
total input and can wrap past `usize::MAX` even while the rolling window
stays bounded. Per-site `saturating_add` is one option but costs a branch
/ clamp on every hot-loop iteration.

This PR hoists the guard once, at the data-ingest boundary:

- `pub(crate) const STREAM_ABS_HEADROOM: usize = HC_OPT_NUM + 16;` —
  slack any downstream `abs_pos + N` site needs.
- `pub(crate) fn check_stream_abs_headroom(history_abs_start,
  window_size, data_len)` — `checked_add` chain that fails fast (with a
  clear panic message pointing 32-bit users at frame splitting or a
  64-bit build) if the next ingest would advance the cursor within
  `STREAM_ABS_HEADROOM` of `usize::MAX`.
- Both backend ingest paths — `MatchTable::add_data` and
  `DfastMatchGenerator::add_data` — route through the same helper.

With that gate in place every per-position `abs_pos + N` site runs as
raw `+` with zero per-iteration overhead.

## What this PR does

**Sites converted from `saturating_*` to raw arithmetic:**

- `bt_insert_step_no_rebase_body!` — `match_end_abs = abs_pos + 9`
- `bt_insert_and_collect_matches_body!` — `match_end_abs = abs_pos + 9`
- `collect_optimal_candidates_initialized_body!` — rep + hash3 + chain
  orchestrator
- `build_optimal_plan_impl_body!` candidate-seed path — `match_end_abs =
  abs_pos + 9`
- `MatchTable` BT walker dispatchers + `encode_offset_with_reps` family
- `DfastMatchGenerator::insert_positions_with_step` — `pos += step` (was
  `pos.saturating_add(step)`)

**Sites migrated to `wrapping_add` (modulo-ring semantics, not overflow
guard):**

- `bt_pair_index_for_abs` — `abs_pos.wrapping_add(self.index_shift)`,
  immediately masked by `& bt_mask()`. The modulo-ring identity keeps
  the release-mode overflow branch and debug-mode panic off this hot
  path even on rare 32-bit overflows.

**Sites intentionally kept as `saturating_*` (underflow at frame start /
EOF — different concern from overflow):**

- `bt_low = abs_pos.saturating_sub(bt_mask)` — fresh-frame case
- `tail_limit = current_abs_end.saturating_sub(abs_pos)` — EOF case
- `*_saturating_sub(BOUNDARY_DENSE_TAIL_LEN)` — dfast tail seeding
- `actual_offset.saturating_add(3)` in `encode_offset_with_reps` —
  external `u32::MAX` input trust boundary

For these, raw subtraction would wrap into a huge `usize` and translate
into OOB reads — the saturating op is a safety guard, not a perf cost.

## Level22 measurements (criterion --quick, x86_64-darwin)

Pure_rust median, before vs after the full series:

| Scenario              | Pre #119 baseline | Post Phase 2 | Δ           |
| ---                   | ---               | ---          | ---         |
| decodecorpus-z000033  | 471.85 ms         | 384.10 ms    | **-18.6 %** |
| high-entropy-1m       | 3.56 ms           | (noisy)      | within noise band |
| low-entropy-1m        | 1.38 ms           | 1.23 ms      | **-10.9 %** |
| large-log-stream      | 131.92 ms         | 114.86 ms    | **-12.9 %** |

Rust vs C FFI gap on decodecorpus-z000033 narrowed from **2.95×** (post
#119 merge) to **2.39×** today. The remaining gap is driven by phases
3-5 of #111 (const-generic Strategy dispatch, BT/HC body specialization,
LDM integration).

Acceptance criterion #120 asked ≥10 %; this PR delivers 10–19 %.

## Verification

- build × 2 (default + `--no-default-features`)
- `cargo clippy --all-targets -p structured-zstd -- -D warnings` clean
  (default + `bench_internals`)
- 451 / 451 nextest, including
  `level22_sequences_match_donor_on_corpus_proxy` ratio gate
- macOS-latest CI workaround: prepend `~/.rustup/toolchains/<tc>/bin`
  to `GITHUB_PATH` (Homebrew rustup ships without `cargo`/`rustc` shims
  under `~/.cargo/bin`)

Part of #111.

## Summary by CodeRabbit

* **Refactor**
  * Replaced saturating arithmetic with plain arithmetic across matching, indexing and encoding paths; added assertions and safety comments to preserve invariants.
* **Bug Fixes**
  * Added stream headroom validation and tightened cursor/index progression to prevent overflow/underflow and incorrect pricing near limits.
* **Tests**
  * Added tests covering wrap‑around and headroom boundary conditions.
* **Chores**
  * CI: improved macOS Rust toolchain invocation for more reliable macOS test runs.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/structured-zstd/pull/121)

### CODE REVIEW

File: zstd/src/encoding/match_generator.rs [hunk 1]
```diff
@@ -1071,9 +1075,16 @@ macro_rules! bt_insert_step_no_rebase_body {
...
+        // a fresh frame (bt_low effectively "no floor"). Saturating keeps
+        // the floor at 0 so the `candidate_abs <= bt_low` check never
+        // triggers early; raw subtraction would underflow into a huge
+        // sentinel that ALWAYS triggers.
         let bt_low = $abs_pos.saturating_sub(bt_mask);
         let window_low = $table.window_low_abs_for_target($target_abs);
-        let mut match_end_abs = $abs_pos.saturating_add(9);
+        // `abs_pos + 9` only overflows on usize::MAX-adjacent input which
+        // never occurs in practice — encode blocks cap at HC_BLOCKSIZE_MAX.
+        let mut match_end_abs = $abs_pos + 9;
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_generator.rs]: @coderabbitai Fixed — replaced the loose 'never overflows in practice' rationale with a concrete `debug_assert!(abs_pos <= usize::MAX - 9)` and a comment proving the bound via the caller's `idx + 8 <= concat.len()` precondition. This now holds on 32-bit i686 as well.

**Comment by polaz** [comment]
@coderabbitai Re: outside-diff suggestion on `match_generator.rs:51-56` — fixed in 5b8724bd; the import cfg is now restricted to `target_arch = "x86_64"`.

File: zstd/src/encoding/match_generator.rs [hunk 2]
```diff
@@ -1327,10 +1351,20 @@ macro_rules! build_optimal_plan_impl_body {
...
+                    // `initial_litlen` is the litlen carried from prior
+                    // optimal-plan segments — its real bound is the
+                    // current block length (the frame compressor caps
+                    // block scan at `HC_BLOCKSIZE_MAX`), not the segment
+                    // `current_len`. `p < min_match_len` (small constant),
+                    // so the seed-position litlen stays well within
+                    // `u32::MAX`; `u32::try_from` panics rather than
+                    // silently truncates if the invariant is ever broken.
+                    nodes[p].litlen = u32::try_from(initial_litlen + p)
+                        .expect("optimal parser seed litlen out of u32 range");
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_generator.rs:1374]: @coderabbitai Fixed — the `usize` addition is now guarded explicitly: `initial_litlen.checked_add(p).and_then(u32::try_from).expect(...)`. The `checked_add` runs before `try_from` so on i686 (32-bit `usize`) an overflow at the addition itself is caught instead of slipping past the conversion.

File: zstd/src/encoding/match_table/storage.rs [hunk 3]
```diff
@@ -450,7 +450,21 @@ impl MatchTable {
...
+        // bounded by the same block scan that called us. Their sum
+        // therefore stays well below `usize::MAX` on every supported
+        // target (i686 included). Enforce the invariant with a
+        // debug-only assertion so release builds run the raw `+` with
+        // zero overhead.
+        debug_assert!(
+            abs_pos.checked_add(self.index_shift).is_some(),
+            "bt_pair_index_for_abs: abs_pos + index_shift overflowed usize"
+        );
+        2 * ((abs_pos + self.index_shift) & self.bt_mask())
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_table/storage.rs]: @coderabbitai Fixed — switched to `abs_pos.wrapping_add(self.index_shift)` as suggested. The result is masked down to `bt_mask()` width anyway, so the modulo semantics of `wrapping_add` produce the same ring slot and the hot path stays branch-free with no debug-mode panic risk on long 32-bit streams.

**Comment by polaz** [comment]
@coderabbitai Re: the duplicate finding on `abs_pos + 9` (`match_generator.rs` 1081-1093 / 2111-2118 / 2361-2368) and the outside-diff finding on the BT walker step cap (`storage.rs` 865-871 + dispatchers): both addressed in commits `d9491e00` and `14a577d4`.

- All three `match_end_abs = abs_pos + 9` sites now use `saturating_add(9)`. The previous "can't process more than `usize::MAX` because of RAM" rationale was wrong for streaming encoders (window memory bounded, stream cursor cumulative); saturating clamps to `usize::MAX` and downstream comparisons treat it as "no further extension", a safe degenerate state at stream tail.
- `replay_history_for_rebase_bt` and the 4 BT walker SIMD dispatchers now cap the per-iteration step at the remaining distance via `forward.max(1).min(abs_pos - pos)`, so the cursor cannot wrap past `abs_pos`.
- The dfast loop comments are rewritten to document the actual upstream contract (frame-compressor total-input cap) instead of the false "RAM bound" claim.

File: zstd/src/encoding/dfast/mod.rs [hunk 4]
```diff
@@ -208,6 +208,27 @@ impl DfastMatchGenerator {
...
+        //    retained window size. A long streaming encode on i686
+        //    can therefore push `current_abs_start + pos` past
+        //    `usize::MAX` even though memory usage stays bounded by
+        //    `window_size`. The runtime enforcement lives in
+        //    `MatchTable::add_data` (see `STREAM_ABS_HEADROOM` =
+        //    `HC_OPT_NUM + 16`): every ingest fails fast with a
+        //    clear panic if cumulative input would push the cursor
+        //    within `STREAM_ABS_HEADROOM` of `usize::MAX`. Raw `+`
+        //    here is donor parity and is correct precisely because
+        //    that upstream gate runs before this loop sees any new
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/dfast/mod.rs]: @coderabbitai Fixed in 4c44cb5d — added `check_stream_abs_headroom` helper in `match_table::storage`, wired it into both `MatchTable::add_data` and `DfastMatchGenerator::add_data`, and rewrote the loop comments to cite the dfast enforcement site directly. Also dropped the misleading `current_len <= HC_BLOCKSIZE_MAX` claim — the real block-local bound is `current_len` itself.

**Comment by polaz** [comment]
@coderabbitai Re: the outside-diff finding "Missing overflow protection that the loop comments depend on" (dfast/mod.rs:109-122) — fixed in 4c44cb5d. Both `MatchTable::add_data` and `DfastMatchGenerator::add_data` now route through a shared `check_stream_abs_headroom` helper (in `match_table::storage`), and the loop-invariant comments at lines 219-238 and 294-313 were rewritten to cite `DfastMatchGenerator::add_data` / `check_stream_abs_headroom` directly. The `HC_BLOCKSIZE_MAX` claim was also dropped — the real block-local bound is `current_len`.
